recbole/trainer/trainer.py

```diff
@@ -621,35 +620,31 @@ def evaluate(
             )
         self.eval_collector.model_collect(self.model)
         struct = self.eval_collector.get_data_struct()
-        result = self.evaluator.evaluate(struct)
         if not self.config["single_spec"]:
-            result = self._map_reduce(result, num_sample)
+            struct = self._gather_evaluation_resources(struct)
+        result = self.evaluator.evaluate(struct)
         self.wandblogger.log_eval_metrics(result, head="eval")
         return result
 
-    def _map_reduce(self, result, num_sample):
-        gather_result = {}
-        total_sample = [
-            torch.zeros(1).to(self.device) for _ in range(self.config["world_size"])
-        ]
-        torch.distributed.all_gather(
-            total_sample, torch.Tensor([num_sample]).to(self.device)
-        )
-        total_sample = torch.cat(total_sample, 0)
-        total_sample = torch.sum(total_sample).item()
-        for key, value in result.items():
-            result[key] = torch.Tensor([value * num_sample]).to(self.device)
-            gather_result[key] = [
-                torch.zeros_like(result[key]).to(self.device)
-                for _ in range(self.config["world_size"])
-            ]
-            torch.distributed.all_gather(gather_result[key], result[key])
-            gather_result[key] = torch.cat(gather_result[key], dim=0)
-            gather_result[key] = round(
-                torch.sum(gather_result[key]).item() / total_sample,
-                self.config["metric_decimal_place"],
-            )
-        return gather_result
+    def _gather_evaluation_resources(self, struct: DataStruct) -> DataStruct:
+        """
+        Gather the evaluation resources from all ranks, e.g., 'rec.items', 'rec.score', 'data.label'
+        Only 'rec.*' and 'data.label' are gathered, because they are distributed into different ranks.
+        Args:
+            struct: data struct collected from all ranks
+    
+        Returns: gathered data struct
+    
+        """
+        gather_struct = DataStruct(struct)
+        for key, value in struct:
+            # Adjust the condition according to
+            # [the key definition in evaluator](/docs/source/developer_guide/customize_metrics.rst#set-metric_need)
+            if key.startswith("rec.") or key == "data.label":
```

Comment:
Only `rec.*` and `data.label` are gathered, because they are distributed into different GPUs. While other `key`s like `data.num_items`, `data.num_users` are the same across different GPUs.

The `key`s refer to [docs/source/developer_guide/customize_metrics.rst#set-metric_need](https://github.com/RUCAIBox/RecBole/blob/master/docs/source/developer_guide/customize_metrics.rst#set-metric_need)